reaper/cluster.go

```diff
@@ -0,0 +1,243 @@
+	NextActivation    time.Time `json:"next_activation,omitempty"`
+}
+
+// All the following types are used internally by the client and not part of the public API
+
+type clusterStatus struct {
+	Name           string     `json:"name"`
+	JmxUsername    string     `json:"jmx_username,omitempty"`
+	JmxPasswordSet bool       `json:"jmx_password_is_set,omitempty"`
+	Seeds          []string   `json:"seed_hosts,omitempty"`
+	NodeStatus     nodeStatus `json:"nodes_status"`
+}
+
+type nodeStatus struct {
+	EndpointStates []gossipStatus `json:"endpointStates,omitempty"`
+}
+
+type gossipStatus struct {
+	SourceNode    string   `json:"sourceNode"`
+	EndpointNames []string `json:"endpointNames,omitempty"`
+	TotalLoad     float64  `json:"totalLoad,omitempty"`
+	Endpoints     map[string]map[string][]endpointStatus
+}
+
+type endpointStatus struct {
+	Endpoint       string  `json:"endpoint"`
+	DataCenter     string  `json:"dc"`
+	Rack           string  `json:"rack"`
+	HostId         string  `json:"hostId"`
+	Status         string  `json:"status"`
+	Severity       float64 `json:"severity"`
+	ReleaseVersion string  `json:"releaseVersion"`
+	Tokens         string  `json:"tokens"`
+	Load           float64 `json:"load"`
+}
+
+func (c *client) GetClusterNames(ctx context.Context) ([]string, error) {
+	res, err := c.doGet(ctx, "/cluster", nil, http.StatusOK)
+	if err == nil {
+		clusterNames := make([]string, 0)
+		err = c.readBodyAsJson(res, &clusterNames)
+		if err == nil {
+			return clusterNames, nil
+		}
+	}
+	return nil, fmt.Errorf("failed to get cluster names: %w", err)
+}
+
+func (c *client) GetCluster(ctx context.Context, name string) (*Cluster, error) {
+	path := "/cluster/" + url.PathEscape(name)
+	res, err := c.doGet(ctx, path, nil, http.StatusOK)
+	if err == nil {
+		clusterStatus := &clusterStatus{}
+		err = c.readBodyAsJson(res, clusterStatus)
+		if err == nil {
+			return newCluster(clusterStatus), nil
+		}
+	}
+	return nil, fmt.Errorf("failed to get cluster %s: %w", name, err)
+}
+
+// Fetches all clusters. This function is async and may return before any or all results are
+// available. The concurrency is currently determined by min(5, NUM_CPUS).
+func (c *client) GetClusters(ctx context.Context) <-chan GetClusterResult {
+	// TODO Make the concurrency configurable
+	concurrency := int(math.Min(5, float64(runtime.NumCPU())))
+	results := make(chan GetClusterResult, concurrency)
+
+	clusterNames, err := c.GetClusterNames(ctx)
+	if err != nil {
+		close(results)
+		return results
+	}
+
+	var wg sync.WaitGroup
+
+	go func() {
+		defer close(results)
+		for _, clusterName := range clusterNames {
+			wg.Add(1)
+			go func(name string) {
+				defer wg.Done()
+				cluster, err := c.GetCluster(ctx, name)
+				result := GetClusterResult{Cluster: cluster, Error: err}
+				results <- result
+			}(clusterName)
+		}
+		wg.Wait()
+	}()
+
+	return results
+}
+
+// Fetches all clusters in a synchronous or blocking manner. Note that this function fails
```

Comment:
One Go commenting standard I learned, not sure if it's essential for us, is to start the comment with the name of the function.